http/src/main/java/io/micronaut/http/filter/FilterRunner.java

```diff
@@ -941,168 +742,244 @@ private DelayedFilterReturnHandler(boolean isResponseFilter, FilterReturnHandler
-     * {@link #workResponse()} in case of errors) is called exactly once. For this purpose, this
-     * class has the {@link #downstreamGuard} flag. It is set to {@link true} when the filter
-     * calls {@link #proceed}, or when the filter throws an exception. Only the first of these
-     * events actually continues working the {@link FilterRunner}, any events that follow are
-     * discarded (logged or throw an exception).
-     *
-     * @param <R> Return value of the continuation
-     */
-    private abstract class FilterContinuationImpl<R> extends SuspensionPoint implements FilterContinuation<R> {
         /**
-         * This flag guards <i>downstream</i> execution, i.e. the call to {@link #workRequest()}
-         * that runs downstream filters. Only the thread that claims this guard may run
-         * {@link #workRequest()}.
+         * The future indicating that the next filter should be executed.
          */
-        final AtomicBoolean downstreamGuard = new AtomicBoolean(false);
+        final CompletableFuture<FilterContext> nextFilterProcessing = new CompletableFuture<>();
         /**
-         * This flag guards <i>upstream</i> execution, i.e. the call to {@link #workResponse()}
-         * after the downstream has returned the response, or if the downstream pipeline was
-         * aborted (usually because of an error).
+         * The future representing the suspension point, completing it will resume this filter processing.
          */
-        final AtomicBoolean upstreamGuard = new AtomicBoolean(false);
+        final CompletableFuture<FilterContext> suspensionPoint = new CompletableFuture<>();
         /**
-         * Executor to run any downstream reactive code on. Only used by some implementations, e.g.
-         * it doesn't make sense for a blocking continuation.
+         * The future representing the filter return value and will be completed when the filter method is finally processed.
          */
-        @Nullable
-        Executor completeOn = null;
+        final CompletableFuture<FilterContext> filterProcessed = new CompletableFuture<>();
 
-        FilterContinuationImpl(int filterIndex, @Nullable SuspensionPoint next) {
-            super(filterIndex, next);
+        FilterContinuationImpl(FilterContext filterContext) {
+            this.filterContext = filterContext;
         }
 
         @Override
         public FilterContinuation<R> request(HttpRequest<?> request) {
-            FilterRunner.this.request = Objects.requireNonNull(request, "request");
+            filterContext = filterContext.withRequest(Objects.requireNonNull(request, "request"));
             return this;
         }
 
-        final void triggerDownstreamWorkRequest() {
-            if (downstreamGuard.compareAndSet(false, true)) {
-                workRequest();
+        protected final void proceedRequested() {
+            if (!nextFilterProcessing.isDone()) {
+                nextFilterProcessing.complete(filterContext);
```

Comment:
As I understand it, nextFilterProcessing is completed by the proceed call, which is then checked later (via `nextFilterFlow()`) to *actually* proceed. Why is this necessary? It resolves some conflicts (e.g. if proceed was called but the method also throws an exception, it will simply not perform the proceed), but other than that it just obfuscates the control flow imo, and it's the reason why the one test had to be ignored

Reply:
The `nextFilterProcessing` will make the `nextFilterFlow` emit the value, which will go to the next filter, etc. And according to the previous code it needs to proceed to the next filter only after the subscription.

The test that was ignored was continuing the flow process in the proceed method, that's why it was working.

Reply:
for reactor it has to proceed only after the subscription, because otherwise we can't propagate the context. But for CompletableFuture and synchronous proceed, it can proceed immediately

Reply:
Yes, the best would be to have reactive/async implementation of the continuation. And for reactive ones, it might be better to use Reactor sinks.